io/mpi/src/TMPIFile.cxx

```diff
@@ -0,0 +1,592 @@
+         // retrieve the message
+         MPI_Recv(buf, number_bytes, MPI_CHAR, source, tag, fSubComm, MPI_STATUS_IGNORE);
+
+         // create a TMemFile from the buffer
+         TMemFile *infile = new TMemFile(fMPIFilename, buf, number_bytes, "UPDATE");
+         if (infile->IsZombie()) {
+            exit(1);
+         }
+         // match compression settings of this TMPIFile object
+         infile->SetCompressionSettings(this->GetCompressionSettings());
+
+         // retrieve existing output file object
+         ParallelFileMerger *info = (ParallelFileMerger *)mergers.FindObject(fMPIFilename);
+         // if exiting file does not exist, create a new one
+         if (!info) {
+            info = new ParallelFileMerger(fMPIFilename, this->GetCompressionSettings(), cache);
+            // add file to hash table
+            mergers.Add(info);
+         }
+
+         // first merge needs extra care
+         if (TMPIFile::R__NeedInitialMerge(infile)) {
+            info->InitialMerge(infile);
+         }
+
+         // merge the data
+         info->RegisterClient(client_Id, infile);
+         info->Merge();
+         infile = 0;
+
+         client_Id++;
+      }
+      delete[] buf;
+   }
+
+   if (fEndProcess == fMPILocalSize - 1) {
+      mergers.Delete();
+      return;
+   }
+}
+
+Bool_t TMPIFile::R__NeedInitialMerge(TDirectory *dir)
+{
+   if (dir == 0)
+      return kFALSE;
+   TIter nextkey(dir->GetListOfKeys());
+   TKey *key;
+   while ((key = (TKey *)nextkey())) {
+      TClass *cl = TClass::GetClass(key->GetClassName());
+      if (cl->InheritsFrom(TDirectory::Class())) {
+         TDirectory *subdir = (TDirectory *)dir->GetList()->FindObject(key->GetName());
+         if (!subdir) {
+            subdir = (TDirectory *)key->ReadObj();
+         }
+         if (TMPIFile::R__NeedInitialMerge(subdir)) {
+            return kTRUE;
+         }
+      } else {
+         if (0 != cl->GetResetAfterMerge()) {
+            return kTRUE;
+         }
+      }
+   }
+   return kFALSE;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// Constructor for ParallelFileMerger class
+
+TMPIFile::ParallelFileMerger::ParallelFileMerger(const char *filename, Int_t compression_settings, Bool_t writeCache)
+   : fFilename(filename), fClientsContact(0), fMerger(kFALSE, kTRUE)
+{
+   fMerger.SetPrintLevel(0);
+   if (!fMerger.OutputFile(filename, "RECREATE"))
+      exit(1);
+   fMerger.GetOutputFile()->SetCompressionSettings(compression_settings);
+   if (writeCache)
+      new TFileCacheWrite(fMerger.GetOutputFile(), 32 * 1024 * 1024);
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// Deconstructor for ParallelFileMerger class
+///
+/// taken from root/tutorials/net/parallelMergeServer.C
+
+TMPIFile::ParallelFileMerger::~ParallelFileMerger()
+{
+   for (ClientColl_t::iterator iter = fClients.begin(); iter != fClients.end(); ++iter)
+      delete iter->GetFile();
```

Comment:
Since `ClientColl_t` (an unnecessary obscured typedef in my opinion) is just an `std::vector`, this loop can be
```
for (auto& client : fClients)
   delete client.GetFile();
```
Similarly for other loops in functions copied from the `parallelMergeServer.C` tutorial. They could use a C++11 face lift before being committed.